### ISSUE TITLE
PostgreSQL already running, but status shown as not.

### ISSUE THREAD

**Issue by klonos** [WPN-XM/WPN-XM#398]
It's been a long time without any issues. The software has matured a great deal. Thank you for all the effort and time you spend on this project.

Despite no major issues, there is a minor annoyance I have for a while now. I thought it was a regression and that perhaps somebody would have already reported it so that it would be fixed in a future version. I've searched the issue queue and I see no issues related to it. Enough with the prologue...

Every time I run the SCP (say after a system reboot) and start all daemons, there is warning saying that PostgreSQL is already running. This is weird because during installation, I did not set anything to autostart as a service (I've checked my autoruns to confirm this). Even if so, shouldn't the service light show as green/on instead of being gray? Anyway, hitting the respective stop button, waiting for a couple of seconds and then hitting the run button restarts the service successfully and this time the status light is shown as green/on.

As I said, not a major thing - only a minor annoyance since there is a workaround or it can simply be ignored.

Shouldn't be OS-specific or system specific, but I'm on Win8.1 x64 running the latest WPN-XM v0.8.2 with the SCP updated to 0.8.3+2cdb580. Let me know if you need any log files or anything in order to troubleshoot this one.

**Comment by jakoch** [comment]
Hey Gregory!

> It's been a long time without any issues. The software has matured a great deal. Thank you for all the effort and time you spend on this project.

Thank you, I appreciate it.

> As I said, not a major thing - only a minor annoyance since there is a workaround or it can simply be ignored.

No, it's better to address the issue. So: thanks for reporting! 

---

The process monitoring for PostgreSQL is PID file based. 
The important file is `/bin/pgsql/data/postmaster.pid`.
If the file is present, then the SCP will return the "process state" has "running".

The PostgreSQL "Stop Button" does an additional PID file reset:
for instance, in case of a failed shutdown the PID file is deleted, to always allow a clean restart.

> Every time I run the SCP (say after a system reboot) and start all daemons, there is warning saying that PostgreSQL is already running.

Are you referring to this dialog: "The following processes are already running:" + checkbox(es) for shutdown or continue? In other words: PostgreSQL is not running, but you get the dialog?

**Comment by klonos** [comment]
I don't know if it is related or not, but when I stop/start all daemons PostgreSQL takes about a second before it starts while the rest of the daemons start instantly. This is reproducible *every time*.

> In other words: PostgreSQL is not running, but you get the dialog?

The issue occurs only after a system restart and to tell you the truth the reason why I did not report the issue earlier is because it takes weeks for me to restart my computer. So, I reported the issue but did not actually see if postgres.exe was running before starting SCP.

If you bare with me and keep this issue open in the meantime, I might restart over the weekend and actually check if postgres.exe runs on startup somehow (I doubt it though because -as I've already said- I checked my startups).

**Comment by jakoch** [comment]
Ok. 
Let connect this to code.

The process lookup at SCP start is done in the function [MainWindow::checkAlreadyActiveDaemons()](https://github.com/WPN-XM/server-control-panel/blob/master/src/mainwindow.cpp#L836).
The process name [searched for](https://github.com/WPN-XM/server-control-panel/blob/master/src/mainwindow.cpp#L867) is "postgres".
If you click continue, when "postgres" is running, the "Status Led" needs an update (it's off by default). 

This is done by calling MainWindow::setLabelStatusActive() from here: https://github.com/WPN-XM/server-control-panel/blob/master/src/mainwindow.cpp#L989

---

> I don't know if it is related or not, but when I stop/start all daemons PostgreSQL takes about a second before it starts while the rest of the daemons start instantly. This is reproducible every time.

Yes, [there is an delay(1250) in there](https://github.com/WPN-XM/server-control-panel/blob/6008371dc4b088035b5a0b41d316084f7a972b54/src/servers.cpp#L722), to wait for the process to start up and for the PID file to be created.

Maybe a for iteration with i=5 and file.exists() plus delay(200) is faster...
I'm open for suggestions to speed this up...

**Comment by klonos** [comment]
I don't mind the delay - just mentioning it in case it is related to the issue at hand. If needed I'll let somebody else file a separate issue for it.

As I said, I suspect that postgres.exe runs at system startup (somehow?) although not configured to (will test and confirm/refute this). If that theory of mine is confirmed, all that remains is to address the issue of the on led in the SCP not being green. If the theory is refuted, all that remains is to address the issue of the SCP thinking that postgres.exe is running when it's not.

Does that seem like a plan?

**Comment by jakoch** [comment]
sounds like a plan

**Comment by klonos** [comment]
I checked this again and again. I monitor all processes running from the installation dir. Upon SCP, only wpn-xm.exe runs + mongo.exe momentarily and then shuts down (I guess this is in order to retrieve mongo version??).

Same thing happens with mongo.exe when I quit the SCP - despite the fact that I have stopped all daemons, mongo.exe fires momentarily and shuts down before wpn-xm.exe is killed.

Now, to the issue at hand: ...with only wpn-xm.exe running, I hit "Start" to start all daemons and the error message about PostgreSQL comes up. Checking the process monitor though reveals all other daemons running **besides** postgres.exe. So, with the process monitor open in the background monitoring the install dir, I press the stop button for the PostgreSQL daemon. Nothing happens of course. Then I press the start button for the PostgreSQL daemon and I see pg_ctl.exe starting momentarily before postgres.exe is launched and then pg_ctl.exe is killed. No error message this time.

Hope this helps.

**Comment by klonos** [comment]
I can make a screencast if you think that it would help. Let me know.

**Comment by jakoch** [comment]
Regarding the mongo issue:

> Same thing happens with mongo.exe when I quit the SCP - despite the fact that I have stopped all daemons, mongo.exe fires momentarily and shuts down before wpn-xm.exe is killed

I think "Stop all daemons on Quit" is active and when you quit the SCP all the stop commands run.
One of these command is stopMongo(), which fires up `mongo.exe`. 

The  [stopMongoDb()()](https://github.com/WPN-XM/server-control-panel/blob/master/src/servers.cpp#L551) command needs to check at the top, if mongo is really running (and return early from the frunction, if not).

Ok, lets add this check.

---
Regarding the postgresql issue:

This sounds like a failed shutdown.
The stopPostgres() command tries to (1) stop postgres and (2) deletes the PID file,
because it might be, that the process is down, but the PID is still there (indicating that it is running, but its not). So, deleting the PID file is basically a Reset. And after that, the start command should work again. 

`pg_ctl.exe start -> postgres.exe start -> pg_ctl.exe stop` is the normal chained start for Postgres.
Its invoked via the control executable and that's why i can't monitor the process directly, when starting. 

I'm not sure, which changes are required for the postgres issue(s).

**Comment by jakoch** [comment]
I can not reproduce this issue.

> Checking the process monitor though reveals all other daemons running besides postgres.exe. So, with the process monitor open in the background monitoring the install dir, I press the stop button for the PostgreSQL daemon. Nothing happens of course.

The only thing i could reproduce is, that the `status.bat` script is not listing postgres.exe.
